cloudstate/cloudstate.go

```diff
@@ -0,0 +1,249 @@
+			return
+		}
+		if err = cs.eventSourcedHandler.registerEntity(ese); err != nil {
+			return
+		}
+		if err = cs.entityDiscoveryResponder.registerEntity(ese, config); err != nil {
+			return
+		}
+	})
+	return
+}
+
+// Run runs the CloudState instance.
+func (cs *CloudState) Run() error {
+	host, ok := os.LookupEnv("HOST")
+	if !ok {
+		return fmt.Errorf("unable to get environment variable \"HOST\"")
+	}
+	port, ok := os.LookupEnv("PORT")
+	if !ok {
+		return fmt.Errorf("unable to get environment variable \"PORT\"")
+	}
+	lis, err := net.Listen("tcp", fmt.Sprintf("%s:%s", host, port))
+	if err != nil {
+		return fmt.Errorf("failed to listen: %v\n", err)
+	}
+	log.Printf("starting grpcServer at: %s:%s", host, port)
+	if e := cs.server.Serve(lis); e != nil {
+		return fmt.Errorf("failed to grpcServer.Serve for: %v", lis)
+	}
+	return nil
+}
+
+// EntityDiscoveryResponder implements the CloudState discovery protocol.
+type EntityDiscoveryResponder struct {
+	fileDescriptorSet *filedescr.FileDescriptorSet
+	entitySpec        *protocol.EntitySpec
+	message           *descriptor.Message
+}
+
+// NewEntityDiscoveryResponder returns a new and initialized EntityDiscoveryResponder.
+func NewEntityDiscoveryResponder(options *Options) *EntityDiscoveryResponder {
+	responder := &EntityDiscoveryResponder{}
+	responder.entitySpec = &protocol.EntitySpec{
+		Entities: make([]*protocol.Entity, 0),
+		ServiceInfo: &protocol.ServiceInfo{
+			ServiceName:           options.ServiceName,
+			ServiceVersion:        options.ServiceVersion,
+			ServiceRuntime:        fmt.Sprintf("%s %s/%s", runtime.Version(), runtime.GOOS, runtime.GOARCH),
+			SupportLibraryName:    SupportLibraryName,
+			SupportLibraryVersion: SupportLibraryVersion,
+		},
+	}
+	responder.fileDescriptorSet = &filedescr.FileDescriptorSet{
+		File: make([]*filedescr.FileDescriptorProto, 0),
+	}
+	return responder
+}
+
+// Discover returns an entity spec for
+func (r *EntityDiscoveryResponder) Discover(c context.Context, pi *protocol.ProxyInfo) (*protocol.EntitySpec, error) {
+	log.Printf("Received discovery call from sidecar [%s w%s] supporting CloudState %v.%v\n",
+		pi.ProxyName,
+		pi.ProxyVersion,
+		pi.ProtocolMajorVersion,
+		pi.ProtocolMinorVersion,
+	)
+	for _, filename := range []string{
+		"google/protobuf/empty.proto",
+		"google/protobuf/any.proto",
+		"google/protobuf/descriptor.proto",
+		"google/api/annotations.proto",
+		"google/api/http.proto",
+		"cloudstate/event_sourced.proto",
+		"cloudstate/entity.proto",
+		"cloudstate/entity_key.proto",
+	} {
+		if err := r.registerFileDescriptorProto(filename); err != nil {
+			return nil, err
+		}
+	}
+	log.Printf("Responding with: %v\n", r.entitySpec.GetServiceInfo())
+	return r.entitySpec, nil
+}
+
+// ReportError logs any user function error reported by the CloudState proxy.
+func (r *EntityDiscoveryResponder) ReportError(c context.Context, fe *protocol.UserFunctionError) (*empty.Empty, error) {
+	log.Printf("ReportError: %v\n", fe)
```

Comment:
Agreed. For the first section see comment above about logging and https://github.com/cloudstateio/cloudstate/issues/130.

Regarding ReportError: I think the spec also should clarify what such an error means regarding. As there is no correlation to a command, reply or failure how should a user function behave getting such an error? https://github.com/cloudstateio/cloudstate/pull/119 has comments regarding this issue.

In this Draft PR documentation for `ReportError` states:
`// Lets a CloudState Proxy inform a User Function about specification violations.`

for which a user of a user function can't do anything useful.

- [ ] let the user register a function where the protocol.UserFunctionError can be routed to.

Reply:
Should we panic in that case? Seems like a spec violation should kill the function if there is no way to handle it gracefully?

Reply:
Thats a good point, we really could.

It would panic without the entity to be known of tho. As the entity should not change its state without a command. The original protobuf says:

```
    // Report an error back to the user function. This will only be invoked to tell the user function
    // that it has done something wrong, eg, violated the protocol, tried to use an entity type that
    // isn't supported, or attempted to forward to an entity that doesn't exist, etc. These messages
    // should be logged clearly for debugging purposes.
```
Some of these use cases might even need some context, otherwise what would the client possibly do from the moment on this userError was reported. Having a spec error sounds like a terminal error.

I think the spec should clarify further in what state a user function/entity would be after such an error. I've added a comment at the corresponding DRAFT PR for the spec and for the moment just to `panic`.

Reply:
As discussed today with @jroper and @viktorklang ReportError should:

- Should be used for informational purposes only—the message MUST be reported to the developer. Not logged as DEBUG. Typically that would mean writing it to STDOUT but it may go somewhere else depending on the language used.
- If stream is closed afterwards, then it should passivate the entity (just as normal)

therefore, if the proxy does not close the stream, the entity will receive further commands/events for the entity. The user (developer) should receive the error and then decides what to do. It is advised to make the error discoverable, most probably through writing it to stderr/stdout.

Reply:
see https://github.com/cloudstateio/go-support/issues/9